---
title: "Reading 43th releases of Human Genome in R"
author: "Ozan"
date: "2023-03-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read GTF file into R

### Use the latest `GENCODE` GTF file
#### disclaimer: this codes are derived from https://davetang.github.io/muse/read_gtf.html 
```{r}
gencode_ver <- 43
my_url <- paste0("https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_",
                 gencode_ver, "/gencode.v", gencode_ver, ".annotation.gtf.gz")
my_gtf <- basename(my_url)

if (!file.exists(paste0("~/home", my_gtf))){
  download.file(url = my_url, destfile = paste0("~/home", my_gtf))
}

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("rtracklayer")  #use rtracklayer to import the GTF file into R
library(rtracklayer)

install.packages("dplyr")
library(dplyr)

# Read in the GTF file
gtf <- read.table("~/home/my_gtf", sep="\t", header=FALSE)

# Remove rows containing NAs
gtf <- gtf[complete.cases(gtf),]


my_obj <- import(paste0("~/home", my_gtf))

if (seqlevels(){}
class(my_obj)


```

```{r}
my_obj
```

### `plyranges` package by Stuart Lee to find out the number of transcripts on each chromosome

```{r}
install.packages("plyranges")   #plyranges package by Stuart Lee to find out 
library(pylranges)              #the number of transcripts on each chromosome

my_obj %>%
  group_by(seqnames)%>%
  summarise(total = n()) %>%
  as.data.frame()

my_obj %>%
  filter(seqnames == "chrM") %>%   #fetching for mitochondrial transcripts
  head()
```

```{r}
my_obj %>%
  group_by(transcript_type) %>%
  summarise(number = n()) %>%
  as.data.frame() -> my_biotypes

my_biotypes %>%
  dplyr::filter(!is.na(transcript_type)) -> my_biotypes

my_biotypes %>%
  arrange(desc(number)) %>%
  dplyr::pull(transcript_type) -> my_order

my_biotypes$transcript_type <- factor(my_biotypes$transcript_type, levels = my_order)

install.packages("ggplot2")
library(ggplot2)

ggplot(my_biotypes, aes(transcript_type, number)) +
  geom_col() +
  theme_bw() +
  scale_y_log10() +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))

```

```{r}
install.packages("Gviz")
install.packages("GenomicFeatures")
library(Gviz)
library(GenomicFeatures)
my_txdb <- makeTxDbFromGFF(paste0("~/home", my_gtf))

my_start <- 7661779 - 5000     # visualisation transcripts near chr17:7661779-7687538
my_end <- 7687538 + 5000
geneTrack <- GeneRegionTrack(my_txdb, chromosome="chr17", from=my_start, to=my_end)

plotTracks(geneTrack, chromosome="chr17", from=my_start, to=my_end, showId=TRUE)
```
